src/processor.rs

```diff
@@ -26,185 +26,275 @@ impl Processor {
 		invoke_signed(
 		    &transfer_to_taker_ix,
 		    &[
-		        pda_temp_token_account.clone(),
-		        takers_token_to_receive_account.clone(),
-		        pda_account.clone(),
-		        token_program.clone(),
+		        pda_temp_token_account_info.clone(),
+		        takers_token_to_receive_account_info.clone(),
+		        pda_account_info.clone(),
+		        token_program_account_info.clone(),
 		    ],
 		    // This parameter is for authority. In this case, the authority is the PDA. BUT instead of passing in the key for PDA itself, we pass in the seeds (&[b"escrow"] and bump_seed), so that we can recalculate the PDA. If the recalculation and the given PDA keys dont' match, then this instruction fails with AuthenticationError
 		    &[&[&b"escrow"[..], &[bump_seed]]], 
 		)?;
 
+		Self::close_pda_and_escrow(
+			pda_temp_token_account_info,
+			token_program_account_info,
+			initializers_main_account_info,
+			pda,
+			bump_seed,
+			pda_account_info,
+			escrow_account_info
+		)
+	}
+
+	/// Cancel can be called after init_escrow. If called after exchange, it's already too late
+	/// since exchange is atomic and tokens have been transferred, so there will be no effect.
+	/// Since tokens haven't actually been transferred from initializer main token account to
+	/// initializer temp token account, we don't need to actually transfer any tokens.
+	/// What we need to do is:
+	/// 1) Close PDA account
+	/// 2) Close escrow info
+	fn process_cancel(account_infos: &[AccountInfo], program_id: &Pubkey) -> ProgramResult {
+		let account_info_iter = &mut account_infos.iter();
+		let initializer_info = next_account_info(account_info_iter)?;
+
+		if !initializer_info.is_signer {
+			return Err(ProgramError::MissingRequiredSignature);
+		}
+
+		let initializer_token_account_info = next_account_info(account_info_iter)?;
+		if *initializer_token_account_info.owner != spl_token::id() {
+			return Err(ProgramError::IncorrectProgramId);
+		}
+
+		let escrow_account_info = next_account_info(account_info_iter)?;
+		let escrow = Escrow::unpack_unchecked(&escrow_account_info.try_borrow_data()?)?;
+		if !escrow.is_initialized() {
+			return Err(ProgramError::UninitializedAccount);
+		}
+
+		let token_program_account_info = next_account_info(account_info_iter)?;
+
+		let pda_temp_token_account_info = next_account_info(account_info_iter)?;
+		let (pda, bump_seed) = Pubkey::find_program_address(&[b"escrow"], program_id);
+
+		let initializers_main_account_info = next_account_info(account_info_iter)?;
+		// Validate Escrow matches instruction
+		if escrow.temp_token_account_pubkey != *pda_temp_token_account_info.key {
+			return Err(ProgramError::InvalidAccountData);
+		}
+		if escrow.initializer_pubkey != *initializers_main_account_info.key {
+			return Err(ProgramError::InvalidAccountData);
+		}
+		if escrow.initializer_token_to_receive_account_pubkey != *initializer_token_account_info.key {
+			return Err(ProgramError::InvalidAccountData);
+		}
```

Comment:
these checks are shared code in both exchange and cancel and hence should probably share code. I'd also combine the 3 ifs into a single one (since they share the same error return).